File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-0.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
    stack<int> st;
    int maxArea = 0;
    int n = heights.size();
    
    for (int i = 0; i < n; i++) {
        // While the current bar is shorter than the bar at the top of the stack, pop the stack
        while (!st.empty() && heights[st.top()] > heights[i]) {
            int h = heights[st.top()];
            st.pop();
            int width = st.empty() ? i : i - st.top() - 1;
            maxArea = max(maxArea, h * width);
        }
        st.push(i);
    }
    
    // Process any remaining bars in the stack
    while (!st.empty()) {
        int h = heights[st.top()];
        st.pop();
        int width = st.empty() ? n : n - st.top() - 1;
        maxArea = max(maxArea, h * width);
    }
    
    return maxArea;
}
};

File: Data Structures & Algorithms/max-water-container/submission-0.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        maxx = 0

        l = 0
        r = len(heights) - 1

        while l < r:
            curr_max = min(heights[l],heights[r]) * (r - l)
            maxx = max(maxx, curr_max)
            if heights[l] < heights[r]:
                l += 1
            else:
                r -= 1
        
        return maxx